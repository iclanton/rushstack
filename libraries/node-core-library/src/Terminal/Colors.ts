// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * @beta
 */
export interface IColorableSequence {
  text: string;
  isEol?: boolean;
  foregroundColor?: ColorValue;
  backgroundColor?: ColorValue;
  textAttributes?: TextAttribute[];
}

export const eolSequence: IColorableSequence = {
  isEol: true
} as IColorableSequence;

/**
 * Colors used with {@link IColorableSequence}.
 * @beta
 */
export enum ColorValue {
  Black,
  Red,
  Green,
  Yellow,
  Blue,
  Magenta,
  Cyan,
  White,
  Gray
}

/**
 * Text styles used with {@link IColorableSequence}.
 * @beta
 */
export enum TextAttribute {
  Bold,
  Dim,
  Underline,
  Blink,
  InvertColor,
  Hidden
}

export enum ConsoleColorCodes {
  BlackForeground = 30,
  RedForeground = 31,
  GreenForeground = 32,
  YellowForeground = 33,
  BlueForeground = 34,
  MagentaForeground = 35,
  CyanForeground = 36,
  WhiteForeground = 37,
  GrayForeground = 90,
  DefaultForeground = 39,

  BlackBackground = 40,
  RedBackground = 41,
  GreenBackground = 42,
  YellowBackground = 43,
  BlueBackground = 44,
  MagentaBackground = 45,
  CyanBackground = 46,
  WhiteBackground = 47,
  GrayBackground = 100,
  DefaultBackground = 49,

  Bold = 1,

  // On Linux, the "BoldOff" code instead causes the text to be double-underlined:
  // https://en.wikipedia.org/wiki/Talk:ANSI_escape_code#SGR_21%E2%80%94%60Bold_off%60_not_widely_supported
  // Use "NormalColorOrIntensity" instead
  // BoldOff = 21,

  Dim = 2,
  NormalColorOrIntensity = 22,
  Underline = 4,
  UnderlineOff = 24,
  Blink = 5,
  BlinkOff = 25,
  InvertColor = 7,
  InvertColorOff = 27,
  Hidden = 8,
  HiddenOff = 28
}

/**
 * The static functions on this class are used to produce colored text
 * for use with the node-core-library terminal.
 *
 * @example
 * terminal.writeLine(Colors.green('Green Text!'), ' ', Colors.blue('Blue Text!'));
 *
 * @beta
 */
export class Colors {
  public static black(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Black
    };
  }

  public static red(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Red
    };
  }

  public static green(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Green
    };
  }

  public static yellow(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Yellow
    };
  }

  public static blue(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Blue
    };
  }

  public static magenta(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Magenta
    };
  }

  public static cyan(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Cyan
    };
  }

  public static white(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.White
    };
  }

  public static gray(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Gray
    };
  }

  public static blackBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Black
    };
  }

  public static redBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Red
    };
  }

  public static greenBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Green
    };
  }

  public static yellowBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Yellow
    };
  }

  public static blueBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Blue
    };
  }

  public static magentaBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Magenta
    };
  }

  public static cyanBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Cyan
    };
  }

  public static whiteBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.White
    };
  }

  public static grayBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Gray
    };
  }

  public static bold(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Bold);
  }

  public static dim(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Dim);
  }

  public static underline(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Underline);
  }

  public static blink(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Blink);
  }

  public static invertColor(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.InvertColor);
  }

  public static hidden(text: string | IColorableSequence): IColorableSequence {
    return Colors._applyTextAttribute(text, TextAttribute.Hidden);
  }

  public static serializeTextSegmentsToLines(
    segments: (string | IColorableSequence)[],
    withColor: boolean
  ): string[] {
    const lines: string[] = [];
    let segmentsToJoin: string[] = [];
    let lastSegmentWasEol: boolean = false;
    for (let i: number = 0; i < segments.length; i++) {
      const segment: IColorableSequence = Colors._normalizeStringOrColorableSequence(segments[i]);
      lastSegmentWasEol = !!segment.isEol;
      if (lastSegmentWasEol) {
        lines.push(segmentsToJoin.join(''));
        segmentsToJoin = [];
      } else {
        if (withColor) {
          const startColorCodes: number[] = [];
          const endColorCodes: number[] = [];
          switch (segment.foregroundColor) {
            case ColorValue.Black: {
              startColorCodes.push(ConsoleColorCodes.BlackForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Red: {
              startColorCodes.push(ConsoleColorCodes.RedForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Green: {
              startColorCodes.push(ConsoleColorCodes.GreenForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Yellow: {
              startColorCodes.push(ConsoleColorCodes.YellowForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Blue: {
              startColorCodes.push(ConsoleColorCodes.BlueForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Magenta: {
              startColorCodes.push(ConsoleColorCodes.MagentaForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Cyan: {
              startColorCodes.push(ConsoleColorCodes.CyanForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.White: {
              startColorCodes.push(ConsoleColorCodes.WhiteForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }

            case ColorValue.Gray: {
              startColorCodes.push(ConsoleColorCodes.GrayForeground);
              endColorCodes.push(ConsoleColorCodes.DefaultForeground);
              break;
            }
          }

          switch (segment.backgroundColor) {
            case ColorValue.Black: {
              startColorCodes.push(ConsoleColorCodes.BlackBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Red: {
              startColorCodes.push(ConsoleColorCodes.RedBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Green: {
              startColorCodes.push(ConsoleColorCodes.GreenBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Yellow: {
              startColorCodes.push(ConsoleColorCodes.YellowBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Blue: {
              startColorCodes.push(ConsoleColorCodes.BlueBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Magenta: {
              startColorCodes.push(ConsoleColorCodes.MagentaBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Cyan: {
              startColorCodes.push(ConsoleColorCodes.CyanBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.White: {
              startColorCodes.push(ConsoleColorCodes.WhiteBackground);
              endColorCodes.push(ConsoleColorCodes.DefaultBackground);
              break;
            }

            case ColorValue.Gray: {
              startColorCodes.push(ConsoleColorCodes.GrayBackground);
              endColorCodes.push(49);
              break;
            }
          }

          if (segment.textAttributes) {
            for (const textAttribute of segment.textAttributes) {
              switch (textAttribute) {
                case TextAttribute.Bold: {
                  startColorCodes.push(ConsoleColorCodes.Bold);
                  endColorCodes.push(ConsoleColorCodes.NormalColorOrIntensity);
                  break;
                }

                case TextAttribute.Dim: {
                  startColorCodes.push(ConsoleColorCodes.Dim);
                  endColorCodes.push(ConsoleColorCodes.NormalColorOrIntensity);
                  break;
                }

                case TextAttribute.Underline: {
                  startColorCodes.push(ConsoleColorCodes.Underline);
                  endColorCodes.push(ConsoleColorCodes.UnderlineOff);
                  break;
                }

                case TextAttribute.Blink: {
                  startColorCodes.push(ConsoleColorCodes.Blink);
                  endColorCodes.push(ConsoleColorCodes.BlinkOff);
                  break;
                }

                case TextAttribute.InvertColor: {
                  startColorCodes.push(ConsoleColorCodes.InvertColor);
                  endColorCodes.push(ConsoleColorCodes.InvertColorOff);
                  break;
                }

                case TextAttribute.Hidden: {
                  startColorCodes.push(ConsoleColorCodes.Hidden);
                  endColorCodes.push(ConsoleColorCodes.HiddenOff);
                  break;
                }
              }
            }
          }

          for (let j: number = 0; j < startColorCodes.length; j++) {
            const code: number = startColorCodes[j];
            segmentsToJoin.push(...['\u001b[', code.toString(), 'm']);
          }

          segmentsToJoin.push(segment.text);

          for (let j: number = endColorCodes.length - 1; j >= 0; j--) {
            const code: number = endColorCodes[j];
            segmentsToJoin.push(...['\u001b[', code.toString(), 'm']);
          }
        } else {
          segmentsToJoin.push(segment.text);
        }
      }
    }

    if (segmentsToJoin.length > 0) {
      lines.push(segmentsToJoin.join(''));
    }

    if (lastSegmentWasEol) {
      lines.push('');
    }

    return lines;
  }

  /**
   * If called with a string, returns the string wrapped in a {@link IColorableSequence}.
   * If called with a {@link IColorableSequence}, returns the {@link IColorableSequence}.
   *
   * @internal
   */
  public static _normalizeStringOrColorableSequence(value: string | IColorableSequence): IColorableSequence {
    if (typeof value === 'string') {
      return {
        text: value
      };
    } else {
      return value;
    }
  }

  private static _applyTextAttribute(
    text: string | IColorableSequence,
    attribute: TextAttribute
  ): IColorableSequence {
    const sequence: IColorableSequence = Colors._normalizeStringOrColorableSequence(text);
    if (!sequence.textAttributes) {
      sequence.textAttributes = [];
    }

    sequence.textAttributes.push(attribute);
    return sequence;
  }
}
