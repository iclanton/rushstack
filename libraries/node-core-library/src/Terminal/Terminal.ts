// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ITerminalProvider, TerminalProviderSeverity } from './ITerminalProvider';
import { IColorableSequence, ColorValue, Colors, eolSequence } from './Colors';
import { ITerminal } from './ITerminal';

/**
 * This class facilitates writing to a console.
 *
 * @beta
 */
export class Terminal implements ITerminal {
  private _providers: Set<ITerminalProvider>;

  public constructor(provider: ITerminalProvider) {
    this._providers = new Set<ITerminalProvider>();
    this._providers.add(provider);
  }

  /**
   * {@inheritdoc ITerminal.registerProvider}
   */
  public registerProvider(provider: ITerminalProvider): void {
    this._providers.add(provider);
  }

  /**
   * {@inheritdoc ITerminal.unregisterProvider}
   */
  public unregisterProvider(provider: ITerminalProvider): void {
    if (this._providers.has(provider)) {
      this._providers.delete(provider);
    }
  }

  /**
   * {@inheritdoc ITerminal.write}
   */
  public write(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.log);
  }

  /**
   * {@inheritdoc ITerminal.writeLine}
   */
  public writeLine(...messageParts: (string | IColorableSequence)[]): void {
    this.write(...messageParts, eolSequence);
  }

  /**
   * {@inheritdoc ITerminal.writeWarning}
   */
  public writeWarning(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      messageParts.map(
        (part): IColorableSequence => ({
          ...Colors._normalizeStringOrColorableSequence(part),
          foregroundColor: ColorValue.Yellow
        })
      ),
      TerminalProviderSeverity.warning
    );
  }

  /**
   * {@inheritdoc ITerminal.writeWarningLine}
   */
  public writeWarningLine(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      [
        ...messageParts.map(
          (part): IColorableSequence => ({
            ...Colors._normalizeStringOrColorableSequence(part),
            foregroundColor: ColorValue.Yellow
          })
        ),
        eolSequence
      ],
      TerminalProviderSeverity.warning
    );
  }

  /**
   * {@inheritdoc ITerminal.writeError}
   */
  public writeError(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      messageParts.map(
        (part): IColorableSequence => ({
          ...Colors._normalizeStringOrColorableSequence(part),
          foregroundColor: ColorValue.Red
        })
      ),
      TerminalProviderSeverity.error
    );
  }

  /**
   * {@inheritdoc ITerminal.writeErrorLine}
   */
  public writeErrorLine(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      [
        ...messageParts.map(
          (part): IColorableSequence => ({
            ...Colors._normalizeStringOrColorableSequence(part),
            foregroundColor: ColorValue.Red
          })
        ),
        eolSequence
      ],
      TerminalProviderSeverity.error
    );
  }

  /**
   * {@inheritdoc ITerminal.writeVerbose}
   */
  public writeVerbose(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.verbose);
  }

  /**
   * {@inheritdoc ITerminal.writeVerboseLine}
   */
  public writeVerboseLine(...messageParts: (string | IColorableSequence)[]): void {
    this.writeVerbose(...messageParts, eolSequence);
  }

  /**
   * {@inheritdoc ITerminal.writeDebug}
   */
  public writeDebug(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.debug);
  }

  /**
   * {@inheritdoc ITerminal.writeDebugLine}
   */
  public writeDebugLine(...messageParts: (string | IColorableSequence)[]): void {
    this.writeDebug(...messageParts, eolSequence);
  }

  private _writeSegmentsToProviders(
    segments: (string | IColorableSequence)[],
    severity: TerminalProviderSeverity
  ): void {
    const withColorText: { [eolChar: string]: string } = {};
    const withoutColorText: { [eolChar: string]: string } = {};
    let withColorLines: string[] | undefined;
    let withoutColorLines: string[] | undefined;

    this._providers.forEach((provider) => {
      const eol: string = provider.eolCharacter;
      let textToWrite: string;
      if (provider.supportsColor) {
        if (!withColorLines) {
          withColorLines = Colors.serializeTextSegmentsToLines(segments, true);
        }

        if (!withColorText[eol]) {
          withColorText[eol] = withColorLines.join(eol);
        }

        textToWrite = withColorText[eol];
      } else {
        if (!withoutColorLines) {
          withoutColorLines = Colors.serializeTextSegmentsToLines(segments, false);
        }

        if (!withoutColorText[eol]) {
          withoutColorText[eol] = withoutColorLines.join(eol);
        }

        textToWrite = withoutColorText[eol];
      }

      provider.write(textToWrite, severity);
    });
  }
}
