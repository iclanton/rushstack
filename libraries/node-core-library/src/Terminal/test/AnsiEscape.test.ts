// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { AnsiEscape } from '../AnsiEscape';
import { Colors } from '../Colors';
import { StringBufferTerminalProvider } from '../StringBufferTerminalProvider';
import { Terminal } from '../Terminal';

describe('AnsiEscape', () => {
  test('calls removeCodes() successfully', () => {
    const terminalProvider: StringBufferTerminalProvider = new StringBufferTerminalProvider(
      true /* supportsColor */
    );
    const terminal: Terminal = new Terminal(terminalProvider);

    terminal.write(
      Colors.red('H'),
      Colors.yellow('e'),
      Colors.green('l'),
      Colors.blue('l'),
      Colors.magenta('o'),
      Colors.cyan(','),
      ' ',
      Colors.bold('w'),
      Colors.dim('o'),
      Colors.invertColor('r'),
      Colors.white('l'),
      Colors.gray('d'),
      Colors.underline('!')
    );

    const coloredInput: string = terminalProvider.getOutput({ normalizeSpecialCharacters: false });
    expect(coloredInput).toMatchSnapshot();
    const decoloredInput: string = AnsiEscape.removeCodes(coloredInput);
    expect(decoloredInput).toMatchInlineSnapshot(`"Hello, world!"`);
  });
});
