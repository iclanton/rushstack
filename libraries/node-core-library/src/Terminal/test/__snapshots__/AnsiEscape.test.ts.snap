// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnsiEscape calls removeCodes() successfully 1`] = `"[31mH[39m[33me[39m[32ml[39m[34ml[39m[35mo[39m[36m,[39m [1mw[22m[2mo[22m[7mr[27m[37ml[39m[90md[39m[4m![24m"`;
